## Problem #203: Squarefree Binomial Coefficients
[Link to Problem](https://projecteuler.net/problem=203)

### Approach

- idea 1
- idea 2

In [2]:
max_rows = 51

pascal = [[1 for j in range(i + 1)] for i in range(max_rows)]

for i in range(1, max_rows):
    for j in range(1, i):
        pascal[i][j] = pascal[i - 1][j - 1] + pascal[i - 1][j]

max_binom = max(max(row) for row in pascal)
limit = int(max_binom ** 0.5)

is_prime = [True] * (limit + 1)
is_prime[0] = is_prime[1] = False
for i in range(2, int(limit ** 0.5) + 1):
    if is_prime[i]:
        for j in range(i * i, limit + 1, i):
            is_prime[j] = False

primes = [i for i, val in enumerate(is_prime) if val]

unique_binom = set()

for row in pascal:
    for x in row:
        square_free = True
        for p in primes:
            if p * p > x:
                break
            if x % (p * p) == 0:
                square_free = False
                break
        if square_free:
            unique_binom.add(x)

print(sum(unique_binom))  

34029210557338


###### Result: **34029210557338** | Execution time: ~4s

### Complexity analysis

The Pascal triangle construction is $O(n^2)$.

The maximum number on the n-th row on the Pascal triangle is of the order $O(\frac{2^n}{\sqrt{n}})$ (had to use a llm for this one).

The sieve runs in $O(\sqrt{\frac{2^n}{\sqrt{n}}}) = O(2^{\frac{n}{2}} n^{-\frac{1}{4}})$

Things are already getting complicated, but basically the program runs fast enough, so that's what matters 🙂.

##### Tags: #pascal-triangle, #binomials, #factorization